x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Phased.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.plan.logical;
+
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.xpack.esql.analysis.Analyzer;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.util.Holder;
+
+import java.util.List;
+
+/**
+ * Marks a {@link LogicalPlan} node as requiring multiple ESQL executions to run.
+ * All logical plans are now run by:
+ * <ol>
+ *     <li>{@link Analyzer analyzing} the entire query</li>
+ *     <li>{@link Phased#extractFirstPhase extracting} the first phase from the
+ *         logical plan</li>
+ *     <li>if there isn't a first phase, run the entire logical plan and return the
+ *         results. you are done.</li>
+ *     <li>if there is first phase, run that</li>
+ *     <li>{@link Phased#applyResultsFromFirstPhase appling} the results from the
+ *         first phase into the logical plan</li>
+ *     <li>start over from step 2 using the new logical plan</li>
+ * </ol>
+ */
+public interface Phased {
+    /**
+     * Return a {@link LogicalPlan} for the first "phase" of this operation.
+     * The result of this phase will be provided to {@link #nextPhase}.
+     */
+    LogicalPlan firstPhase();
+
+    /**
+     * Use the results of plan provided from {@link #firstPhase} to produce the
+     * next phase of the query.
+     */
+    LogicalPlan nextPhase(List<Attribute> schema, List<Page> firstPhaseResult);
+
+    /**
+     * Find the first {@link Phased} operation and return it's {@link #firstPhase}.
+     * Or {@code null} if there aren't any {@linkplain Phased} operations.
+     */
+    static LogicalPlan extractFirstPhase(LogicalPlan plan) {
+        assert plan.analyzed();
+        var firstPhase = new Holder<LogicalPlan>();
+        plan.forEachUp(t -> {
+            if (firstPhase.get() == null && t instanceof Phased phased) {
+                firstPhase.set(phased.firstPhase());
+            }
+        });
```

Comment:
You can also use `forEachUp` with a class token.
```suggestion
        plan.forEachUp(Phased.class, t -> {
            if (firstPhase.get() == null) {
                firstPhase.set(t.firstPhase());
            }
        });
```
Also applies below.

Reply:
That doesn't work here because `Phased` is an interface and not a subclass of `Node`. I'd have to fiddle with the signature of `forEachUp` in ways with a pretty wide blast radius....

Reply:
Ah, you're right, sorry!